core/cog-modules.c

```diff
@@ -0,0 +1,156 @@
+/*
+ * cog-modules.c
+ * Copyright (C) 2021 Adrian Perez de Castro <aperez@igalia.com>
+ *
+ * Distributed under terms of the MIT license.
+ */
+
+#include "cog-modules.h"
+
+struct ExtensionPoints {
+    GIOExtensionPoint *platform;
+};
+
+static void *
+ensure_extension_points_internal(void *func)
+{
+    static struct ExtensionPoints ep = {};
+    ep.platform = g_io_extension_point_register(COG_MODULES_PLATFORM_EXTENSION_POINT);
+    g_debug("%s: Extension points registered.", (const char *) func);
+    return &ep;
+}
+
+const struct ExtensionPoints *
+ensure_extension_points(void)
+{
+    static GOnce once = G_ONCE_INIT;
+    g_once(&once, ensure_extension_points_internal, (void *) G_STRFUNC);
+    return once.retval;
+}
+
+/**
+ * cog_modules_get_platform_extension_point:
+ *
+ * Obtain the extension point for platform modules.
+ *
+ * Returns: (transfer none): Extension point.
+ */
+GIOExtensionPoint *
+cog_modules_get_platform_extension_point(void)
+{
+    return ensure_extension_points()->platform;
+}
+
+static gboolean
+can_use_extension(GIOExtension *extension, size_t is_supported_offset)
+{
+    if (extension == NULL)
+        return FALSE;
+
+    if (is_supported_offset == 0)
+        return TRUE;
+
+    typedef gboolean (*VerifyFunction)(void);
+
+    GType type = g_io_extension_get_type(extension);
+    g_autoptr(GTypeClass) type_class = g_type_class_ref(type);
+    return (*G_STRUCT_MEMBER(VerifyFunction, type_class, is_supported_offset))();
+}
+
+/**
+ * cog_modules_get_preferred:
+ * @extension_point: Name of the extension point for which to obtain a module.
+ * @preferred_module: (nullable): Name of the preferred module, which
+ *    overrides the default preferred one.
+ * @is_supported_offset: A vtable offset, or zero.
+ *
+ * Retrieves the default class which implements a certain extension point.
+ *
+ * If @preferred_module is not %NULL, then the implementation named by it will
+ * be tried first. After that, or if %NULL, all other implementations will be
+ * tried in order of decreasing priority.
+ *
+ * If @is_supported_offset is non-zero, then it is the offset into the class
+ * vtable at which there is a function which takes no arguments and returns
+ * a boolean. This function will be called on each candidate implementation
+ * to check whether it is usable or not.
+ *
+ * Returns: (transfer none): A type that can be instantiated to implement
+ *    @extension_point, or %G_TYPE_INVALID if there are o usable
+ *    implementations.
+ */
+GType
+cog_modules_get_preferred(GIOExtensionPoint *extension_point, const char *preferred_module, size_t is_supported_offset)
```

Comment:
Convention for GIO modules in GLib is that they have an env var to override this.

See `_g_io_module_get_default_type()`.

Reply:
GIO also adds `GIO_EXTRA_MODULES` so users can specify an extra directory to add to the search path. Similar could be useful.

Reply:
Yes, that's the idea here.

A program using the `libcogcore` library can still choose to read an environment variable, if desired.